Eds 5.0.0

htot opened the pull request:
These changes are targeting Edison with linux 5.0.0 built with Yocto Thud.
1. provides a more recent acpica then OE-Core master (20180508).
2. is generic and needed for sumo and later, otherwise errors are generated
3. and 
4. make the control of the defines through a Yocto variable, prevents needing to edit the asl
5. and 
6. provides 2 more sample tables
7. the recipe as is generates a cpio, this change generates a Package as well containing load script and systemd service for use with configfs. Package can be added to image or initramfs as usual (or not if not needed)
8. Eds needs TRI_STATE_ALL in the load script. If we want this script usable on other platforms we will need some conditional here. Also, this creates a dependency on libgpiod.

htot committed:
acpi-tables: build package with load script and systemd service

This installs a script acpi-tables-load that can be manually run to load all acpi tables in /kernel/firmware/acpi
on a configfs enabled kernel. The script assumes configfs is mounted on /sys/kernel/config as is the default.
Further a systemd service acpi-tables-load.service is installed that runs the script . To run once manually,
use 'systemctl start acpi-tables-load', to automatically run after boot 'systemctl enable acpi-tables-load'

Signed-off-by: Ferry Toth <ftoth@exalondelft.nl>

htot committed:
acpi-tables: control defines through ACPI_FEATURES

This creates a variable ACPI_FEATURES which set defines for uart_2w uart_4w spi and i2c.
This can be used in the machines conf file to control defines while building tables.

Signed-off-by: Ferry Toth <ftoth@exalondelft.nl>

alext-mkrs committed:
meta-acpi: add LAYERSERIES_COMPAT assignments

Starting with Yocto sumo a warning is shown when the compat statement is not there.

Signed-off-by: Alex Tereschenko <alext.mkrs@gmail.com>

htot committed:
acpica: update to 20190215

The recipe was updated to the latest one from openembedded-core.
Acpica version was updated to the latest release available, and
patche for musl dropped (not supported by us), patch for flex
dropped (not needed for thud, not neded for flex 2.6.4 as tested
on Ubuntu).

Signed-off-by: Ferry Toth <ftoth@exalondelft.nl>

htot committed:
acpi-tables: edison: add leds and  ds2-led example

Based on leds.asl for the Minnowboard Max by Mika Westerberg <mika.westerberg@linux.intel.com>

This adds 2 tables
- ds2-led.asl, a solution example of the LED on Edison/Arduino,
- leds.asl an example providing the LED connected to the Edison by the GPIO40 line.

The user can chainload arduino.aml followed by leds.aml or directly load ds2-led.aml.

Note that as ds2-led.aml includes arduino.amli it is affected by ACPI_FEATURES.

Suggested by: Andy Shevchenko <andriy.shevchenko@linux.intel.com>
Signed-off-by: Ferry Toth <ftoth@exalondelft.nl>

andy-shev reviewed (changes requested):
Update version is okay for me, but I don't see the point of the addition patches.

andy-shev reviewed (commented):
I wouldn't apply this till we will be sure that new property makes Linux kernel upstream.

andy-shev reviewed (changes requested):
This needs to be cleaned up. It's not Minnowboard we are targeting here, and is it really a D2 on Edison/Arduino?

andy-shev reviewed (approved):
LGTM!

andy-shev commented on recipes-bsp/acpi-tables/files/acpi-tables-load:
```diff
@@ -3,8 +3,10 @@
-	dest_dir=$(basename $TABLE .aml)
-	mkdir /sys/kernel/config/acpi/table/$dest_dir
-	cat /kernel/firmware/acpi/$TABLE > /sys/kernel/config/acpi/table/$dest_dir/aml
+        dest_dir=$(basename $TABLE .aml)
```
Is this tab vs. space changes? Please, drop.

htot replied:
> Is this tab vs. space changes? Please, drop.
Yes

andy-shev replied:
OK, so, do them in separate change if needed.

andy-shev commented on recipes-bsp/acpi-tables/acpi-tables.bb:
```diff
@@ -14,6 +14,13 @@ inherit deploy
 ACPI_TABLES ?= ""
 ACPI_TABLES[doc] = "List of ACPI tables to include with the initrd"
+ACPI_FEATURES ?= "uart_2w spi i2c"
+IASLCC = " \
```
Sounds like IASLFLAGS or IASLDEFS rather then CC.

htot replied:
> Sounds like IASLFLAGS or IASLDEFS rather then CC.

IASLFLAGS

andy-shev replied:
Good for me.

andy-shev commented on recipes-bsp/acpi-tables/files/acpi-tables-load:
```diff
@@ -3,8 +3,10 @@
+        cat /kernel/firmware/acpi/$TABLE > /sys/kernel/config/acpi/table/$dest_dir/aml
     done
 
+    gpioset `gpiofind TRI_STATE_ALL`=0
```
This is Edison specific in the generic code. NAK.

htot replied:
> This is Edison specific in the generic code. NAK.

I found a way how to do this. Eds gets a specific script, overriding the generic one. When building edison an eds specific package will be built (under edison-poky-linux). When building other machines will be under {arch}-poky-linux. This is behavior as for linux and u-boot.

andy-shev replied:
Fine by me.

andy-shev commented on recipes-bsp/acpi-tables/acpi-tables.bb:
```diff
@@ -20,6 +20,8 @@ B = "${WORKDIR}/acpi-tables"
 
 inherit deploy
 
+RDEPENDS_${PN} = "libgpiod"
```
It's good to have for all of them, indeed.

htot replied:
> It's good to have for all of them, indeed.

Except rdepends on libgpiod maybe. This causes libgpiod package to be installed when acpi-tables is installed. Maybe I should make rdepends conditional on MACHINE = edison?

andy-shev replied:
I think it's still good to have for all boards. But proceed as you consider better.

andy-shev reviewed (changes requested):
Only libgpiod dependency makes sense here.

andy-shev commented on recipes-bsp/acpi-tables/acpi-tables.bb:
```diff
@@ -14,6 +14,13 @@ inherit deploy
+    ${@bb.utils.contains('ACPI_FEATURES', 'uart_2w', '-D MUX_UART_2WIRE', '', d)} \
+    ${@bb.utils.contains('ACPI_FEATURES', 'uart_4w', '-D MUX_UART_4WIRE', '', d)} \
+    ${@bb.utils.contains('ACPI_FEATURES', 'i2c', '-D MUX_I2C', '', d)} \
+    ${@bb.utils.contains('ACPI_FEATURES', 'spi', '-D MUX_SPI', '', d)} \
```
This should also include generic MUX_UART case, I suppose (for non-Edison boards).

htot replied:
> This should also include generic MUX_UART case, I suppose (for non-Edison boards).

Not sure. Search for 'define' on all asl gives only /galileo/gpiomux-*.asl with:
    #define MUX_SPI
    #define MUX_I2C
    #define MUX_UART0

On galileo we are generating a table for each case. gpiomux.asl doesn't define anything which probably should stay that way by default.
So imho I should define MUX_UART0 + not initialize  ACPI_FEATURES ?= "uart_2w spi i2c", or only for edison.

andy-shev replied:
Ah, I see. Than it's fine. we might add an alias to Galileo case that would work as uart_2w and / or uart_4w.

htot replied:
If the user has machine = galileo, and ACPI_TABLES ?= "gpiomux.asl", and ACPI_FEATURES ?= "spi i2c uart0" he will get exactly the same as with ACPI_TABLES ?= "gpiomux-all.asl"

andy-shev replied:
I see. Thanks for explanation. Then  proceed as needed,

andy-shev commented on recipes-bsp/acpi-tables/files/acpi-tables-load:
```diff
@@ -0,0 +1,10 @@
+# This script copies all ACPI tables into /kernel/firmware/acpi
+# You must have configfs enabled for this to work
+
+    for TABLE in `ls /kernel/firmware/acpi/ 2>/dev/null`; do
```
Beeter to use `find /... -type f -name "*.aml"`.

htot replied:
> Beeter to use `find /... -type f -name "*.aml"`.

Maybe leave it for now? Currently we can select which tables are built (ACPI_TABLES). But I can image where 2 tables are built and only one needs to be loaded by default. Like where LED's flash during configuration, and SPI is needed during normal operation. And we could unload SPI and load LED when entering configuration.

In that case we would need an additional configuration file, of pass list of tables to be loaded by doing `sed` on this script.

andy-shev replied:
I meant instead of `ls ...` better to use `$(find ...)`.

htot replied:
Ah, then I will modify.

andy-shev commented on recipes-bsp/acpi-tables/files/acpi-tables-load:
```diff
@@ -0,0 +1,10 @@
+
+    for TABLE in `ls /kernel/firmware/acpi/ 2>/dev/null`; do
+	dest_dir=$(basename $TABLE .aml)
+	mkdir /sys/kernel/config/acpi/table/$dest_dir
```
Maybe `CONFIGFS_DIR="/sys/..."` and then `mkdir $CONFIGFS_DIR/...` and so on?

htot replied:
Good

andy-shev reviewed (changes requested):
Some minor comments, otherwise I like this!

andy-shev reviewed (changes requested):
No, instead simple provide a new file, called *arduino-no-mux.asl*.

htot commented:
> Update version is okay for me, but I don't see the point of the addition patches.

I took acpica recipe from [here](http://cgit.openembedded.org/openembedded-core/tree/meta/recipes-extended/acpica) and upgraded. From the comments it seems like 1 patch needed depending on version of flex and the other when linking against musl.
Eventually OE will update to something newer (and the latest is automatically used in Yocto) and we can drop this recipe. Or go crazy and upstream this one.

htot commented:
> I wouldn't apply this till we will be sure that new property makes Linux kernel upstream.

This relates to `acpi-tables: edison: add sc16is7xx example`? I didn't realize it depends on kernel patch, let me drop this for now.

htot commented:
> This needs to be cleaned up. It's not Minnowboard we are targeting here, and is it really a D2 on Edison/Arduino?

This relates to `acpi-tables: add leds.asl`? I'll clean and push revised patch set.

htot commented:
> No, instead simple provide a new file, called _arduino-no-mux.asl_.

Related to `arduino.asl: remove defines`? I see. Shouldn't we follow galileo naming example:
```
ls gpiomux-*.asl
gpiomux-all.asl  gpiomux.asl  gpiomux-i2c.asl  gpiomux-spi.asl  gpiomux-uart0.asl
```
If so, `gpiomux.asl` has nothing enabled, `gpiomux-all.asl` has everything enabled.

htot commented:
If force pushed v2 taking care of:
- edison specific code is only run when machine = edison, libgpiod dependency created only for edison
- sc16is7xx example dropped for now
- leds example cleaned up

Not yet resolved:
- patches needed for iasl kept. My proposal is keep since these are introduced by OE and probably needed in certain use cases
- arduino.asl: remove defines, your request is to create `arduino-no-mux.asl` which is fine by me. Alternatively, we could create arduino-all.asl and have arduino.asl without includes, following galileo.

I hope I didn't miss anything else. GH seems to not always attach a change request to a particular file or code line, making it hard to track.

andy-shev commented on recipes-bsp/acpi-tables/samples/edison/leds.asl:
```diff
@@ -0,0 +1,88 @@
+/*
+ * Intel Edison
+ *
+ * This table is modified from the Minnowboard Turbot leds.asl for Edison.
```
Here Turbot is mentioned but in the commit message you put MAX. Which one is correct?

htot replied:
Good question. I took from samples/minnowboard-max/leds.asl which in the first line has:
* http://wiki.minnowboard.org/MinnowBoard_Turbot

To be honest I don't know the difference between Minnowboard/Max/Turbot. But I guess this should really be Max?

andy-shev replied:
Minnoboard = v1
... MAX = v2
... Turbot = v3

AFAIU this.

Okay, just check if turbot has that led and use it, or instead you may refer to the file name relative to the source root.

andy-shev commented on recipes-bsp/acpi-tables/samples/edison/leds.asl:
```diff
@@ -0,0 +1,88 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+DefinitionBlock ("leds.aml", "SSDT", 5, "", "LEDS", 1)
```
First of all, since it's an existing LED on Edison/Arduino we are referring to here, I would rather to split this to *leds.asl*:

    DefinitionBlock (...)
    {
      #include "arduino.asli"
      #include "ds2-led.asli"
    }

and move the rest to *ds2-led.asli*.

Otherwise you need to remove all mentions about Arduino and tell about GPIO pin numbering as per 70-pin Edison header.

htot replied:
So how is that supposed to work then? Right now I just load arduino.aml followed by leds.aml which is nice and modular. And then I could unload leds and load spidev.

Maybe I don't understand yet how this is supposed to work?

ds2-led.asl would be a good name though, leds.asl would more fit a table for multiple leds

andy-shev replied:
The _*.asli_ files are the build blocks for "solution" _*.aml_. From the user prospective there is another layer, i.e. when we have _*.aml_ which can be chained (loaded sequentially). Thus we have to provide here (yes, seems above I was a bit wrong) _leds.asl**i**_ (note **i**), and provide *ds2-led.asl* (which would be considered as an example of the LED on Edison/Arduino) and, if you want to, *leds.asl* (note no **i**) as a generic example of the LED connected to the Edison by corresponding GPIO line.

So, user might load *arduino.aml* followed by *leds.aml* or directly load *ds2-led.aml* instead.

P.S. I might have missed something, so, please, double check if it's going to work.

htot replied:
This should be OK, but I will test. However, while we are at it, there is another issue...

andy-shev commented on recipes-bsp/acpi-tables/files/edison/acpi-tables-load:
```diff
@@ -0,0 +1,12 @@
+        cat /kernel/firmware/acpi/$TABLE > /sys/kernel/config/acpi/table/$dest_dir/aml
+    done
+
+    gpioset `gpiofind TRI_STATE_ALL`=0
```
The problem here that you are setting it **after** loading tables. Moreover, the pin name is not known before loading GPIO expanders. As far as I understand there is no solution for the issue, except exporting the pin and setting it inside *arduino.asl* to "high". The workaround is to create a *gpioexp.asl* without any muxing and do something like

- load *gpioexp*
- set TRI_STATE_ALL to "low"
- load tables
- set TRI_STATE_ALL to "high"

But this is fragile since it relies on current behave of GPIO subsystem (leaving GPIO in the last state, there is a discussion to force them to the state before GPIO request had been done).

htot replied:
According to intel docs you need to clear TRI_STATE_ALL, then load tables, then set TRI_STATE_ALL. And as you correctly note TRI_STATE_ALL does not exist before loading the table.

Nevertheless toggling TRI_STATE_ALL is necessary for the GPIO's to work.

I found the order in the script to work. So load tables, clear and set TRI_STATE_ALL. There might be glitches on the GPIO's, I didn't check with the scope. But I think they aren't glitch free on startup anyway.

andy-shev replied:
Yes, the problem is that: if TRI_STATE_ALL is not high already, the loading tables would get non-working drivers (because ->probe() functions fail). I saw this many times.  So above flow I mention wouldn't work either. There is no solution at all, because when tables are loaded via ACPI there is no way to defer driver ->probe(). And there is a HW design issue (who the heck invented TRI_STATE_GPIO to be on the same board as all devices we have to gate?!).

In your case it's "working" because you are actually didn't try complex configurations like Adafruit TFT display.

andy-shev replied:
OK, the order of scanning is defined somehow (but I'm not sure if it's specified, of course it is for parent-child, but the question about siblings), that's why I put *arduino,asli* as a first part in the resulting _*.asl_. The only way to solve this as I can see is to assert TRI_STATE_ALL in U-Boot and then it wouldn't be different how tables are come into place. The drawback is that hardware will have all possible states on the pins.

andy-shev replied:
More I think about it, I come to the following work around:
- specify two identical ASL code with MUXing only (no actual devices described), except one has TRI_STATE_ALL deasserted and the other one asserted.
- specify ASL with the same MUXing, but w/o hogging TRI_STATE_ALL
- do the following: 
 1. Load MUX with TRI_STATE_ALL=0 / unload
 2. Load MUX with TRI_STATE_ALL=1 / unload
 3. Load normal tables w/o TRI_STATE_ALL hogged.

This would guarantee no glitches on the lines AFAICS.

htot replied:
> Yes, the problem is that: if TRI_STATE_ALL is not high already, the loading tables would get non-working drivers (because ->probe() functions fail). I saw this many times. So above flow I mention wouldn't work either. There is no solution at all, because when tables are loaded via ACPI there is no way to defer driver ->probe(). And there is a HW design issue (who the heck invented TRI_STATE_GPIO to be on the same board as all devices we have to gate?!).
> 
> In your case it's "working" because you are actually didn't try complex configurations like Adafruit TFT display.

I thought TRI_STATE_ALL only controls the level translators by setting INH on U5, U7 .. U15 and pulling V_SHIELD_SW low U4, U6 .. U14 and U18 .. U33. But I never really understood how that affects functioning of the muxes. And problems with the muxes after reboot, I haven't seen for a long time, but maybe because I press the RESET longer then before.

Is there no way to define a method in asl that toggles TRI_STATE_ALL on loading the table? Loading/unloading a table sounds like a lot of action. Also, should the correct order not be:
1, 3, 2 (TRI_STATE_ALL goes high after tables loaded)?

andy-shev replied:
It affects devices to be recognized. If we have TRI STATE asserted, it means all devices are virtually disconnected and `->probe()` will fail.

The order I put it sounds correct to me. Becaues only at last stage we load actual devices with MUX tables.

And there is no hooks to do anything on loading or unloading tables. This entire concept (of **un**loading tables) is a hack from ACPI(CA) prospective.

htot replied:
1. I still haven't figured acpi out. But I read there are notify methods for hot insertion and hot removal? And _INI method (as they write: load when OSPM loads a table).

2. 'Devices to be recognized' you mean on the I2C-6 or SPI bus? I am beginning to understand that the Edison HW guide requires toggling TRI STATE only to reset the devices on these busses after configuring external muxes messes up the bus state. Or actually they want to de-assert during configuring external muxes and assert after. But they are only using MRAA from user space so no ->probe() fails.

3. Your are suggesting TRI_STATE_ALL remains unchanged during warm boot. I did some testing (click on summary  to see results):

 <details>
  <summary>Test results assert/de-assert TRI_STATE_ALL</summary>
*** Boot without tables loaded

```
edison login: root
Last login: Sun Mar 24 19:50:39 UTC 2019 on ttyS2
root@edison:~# modprobe -i i2c-dev
root@edison:~# i2cdetect -r 1
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
...
20: 20 21 22 23 -- -- -- -- -- -- -- -- -- -- -- -- 
```

-> not probed

*** Load table

```
root@edison:~# mkdir /sys/kernel/config/acpi/table/arduino
root@edison:~# cat /kernel/firmware/acpi/arduino.aml > /sys/kernel/config/acpi/table/arduino/aml
root@edison:~# i2cdetect -r 1
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
...
20: UU UU UU UU -- -- -- -- -- -- -- -- -- -- -- -- 

root@edison:~# i2cdetect -r 6
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
...
30: -- -- -- -- -- -- -- -- -- 39 -- -- -- -- -- -- 
```

-> muxes probed, device 39 on I2C-6 detected

*** Deassert TRI_STATE_ALL

```
root@edison:~# gpioset `gpiofind TRI_STATE_ALL`=0
root@edison:~# i2cdetect -r 6
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
...
30: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 

root@edison:~# i2cdetect -r 1
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:          -- -- -- -- -- -- -- -- -- -- -- -- -- 
10: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
20: UU UU UU UU -- -- -- -- -- -- -- -- -- -- -- -- 
```

-> mux still probed, device 39 on I2C-6 disappeared (not powered)

*** Reboot with de-asserted TRI_STATE_ALL

```
root@edison:~# reboot usb
edison login: root
root@edison:~# modprobe -i i2c-dev
root@edison:~# i2cdetect -r 1
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:          -- -- -- -- -- -- -- -- -- -- -- -- -- 
10: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
20: 20 21 22 23 -- -- -- -- -- -- -- -- -- -- -- -- 

root@edison:~# i2cdetect -r 6

     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
...
30: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
```

-> device 39 on I2C-6 disappeared (not powered), TRI_STATE_ALL remains de-asserted across reboot

*** Load table

```
root@edison:~# mkdir /sys/kernel/config/acpi/table/arduino
root@edison:~# cat /kernel/firmware/acpi/arduino.aml > /sys/kernel/config/acpi/table/arduino/aml
root@edison:~# i2cdetect -r 1
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
...
20: UU UU UU UU -- -- -- -- -- -- -- -- -- -- -- -- 

root@edison:~# i2cdetect -r 6
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
...
30: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
```

*** Assert TRI_STATE_AL

```
root@edison:~# gpioset `gpiofind TRI_STATE_ALL`=1
root@edison:~# i2cdetect -r 6
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
...
30: -- -- -- -- -- -- -- -- -- 39 -- -- -- -- -- -- 
```

</details> 

These tests show you are right, TRI_STATE_ALL remains de-asserted across reboot.

So, what we should be doing is (note question mark at the end):
- load arduino.aml
- toggle TRI_STATE_ALL
- load tables for i2c-6/spi devices

or use _INI method in i2c-6/spi devices tables to assert TRI_STATE_ALL to power the devices prior to probe()?

andy-shev replied:
1/3. If `_INI()` works it would be considered as a workaround, indeed. Let me try at some point with my devices. It would need to add an Operation region I suppose to the NIO4, where this pin is stated. OTOH I'm not sure if pin would be available to the manual toggling in the way if it's described in OpRegion.
2. Devices connected to the Arduino board. Like I2C sensors.

htot replied:
Test above was with I2C sensor on i2c-6 bus. Except no kernel driver.

Given we drop the patch for now I guess I can mark this as resolved.

andy-shev commented on recipes-bsp/acpica/files/manipulate-fds-instead-of-FILE.patch:
```diff
@@ -0,0 +1,71 @@
+From 540d80469e6a7dce6baf7214df90e86daffc5175 Mon Sep 17 00:00:00 2001
```
No, let's drop these hacks. Update is okay, hacks are not okay for meta-acpi.

htot replied:
I can try, but 'flex 2.6.3 no longer does that, leading to a compiler warning
and link error about yy_scan_string()."

Weird thing is sumo/thud/master have flex 2.6.0. So should be good. Unless they try to build against host (Ubuntu 18.10 has 2.6.4).

andy-shev replied:
I see, so, the patch should be upstreamed then.

htot replied:
The patches are marked as:

> Upstream-Status: Inappropriate [embedded specific]\
> Signed-off-by: Patrick Ohly <patrick.ohly@intel.com>
>
However, I upgraded now to 20190215, tested building on Thud with the patches dropped. Also tested native on Ubuntu 18.10 (flex 2.6.4 ans bison 3.0.4) with no errors. 20190215 has acpi 6.3 and is in kernel 5.1-rc1.
This will appear in v3 of this PR.

andy-shev commented on recipes-bsp/acpi-tables/samples/edison/arduino.asl:
```diff
@@ -23,9 +23,5 @@
  */
 DefinitionBlock ("arduino.aml", "SSDT", 5, "", "ARDUINO", 1)
 {
-    #define MUX_I2C
```
No, if you feel this is a need, split either by scheme I proposed, or one used in Galileo

htot replied:
I think alternative is to create predefined asl's, one for each case (3 defines = 8 cases).

If you ask me, I would prefer to split out and have a mux.asl, uart.asl, uart4w.asl, spi.asl etc. and load the combination needed for a specific setup at run time. I think there is a 256 limit on SSDT's we can load?

Otherwise, having a single asl, with defines controlled from the machine.conf file (so even outside of recipe) is the next best thing, see [machine/edison.conf](https://github.com/htot/meta-intel-edison/blob/master/meta-intel-edison-bsp/conf/machine/edison.conf) last 2 lines.

andy-shev replied:
But this can be created with one ASL source. Just provide as many defines through command line as you want. So, basically we need arduino-no-mux asl (or arduino-all -- whatever you prefer) and Makefile to handle proposed variants.

htot replied:
Yes, using outside of Yocto, with a Makefile would work nicely.
Currently recipe acts as simple Makefile i.e. deletes all targets and build them all new. The recipe would be prettier if it would also call the Makefile. But I'm not really a Makefile wizard and didn't want to touch the recipe too much when I started this.
With a Makefile we could eleminate also from the other platforms unneeded asl's.

htot replied:
For now created arduino-all and leave makefile for future improvement.

andy-shev commented:
> Not yet resolved:
> 
> * patches needed for iasl kept. My proposal is keep since these are introduced by OE and probably needed in certain use cases

meta-acpi should be kept hack-clean repo.

> * arduino.asl: remove defines, your request is to create `arduino-no-mux.asl` which is fine by me. Alternatively, we could create arduino-all.asl and have arduino.asl without includes, following galileo.

Yes, use alternative or scheme I proposed earlier.

> I hope I didn't miss anything else. GH seems to not always attach a change request to a particular file or code line, making it hard to track.

Yeah, it also changes text depending on code context.

htot commented on recipes-bsp/acpi-tables/samples/edison/leds.asl:
```diff
@@ -0,0 +1,88 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+DefinitionBlock ("leds.aml", "SSDT", 5, "", "LEDS", 1)
```
... The LED is connected to GP40, but not really. It's connected to DIG13_1P8. And this is connected to either GP40 or GP109 depending on SPI_CLK_SEL (U16 IO1.3). So when we enable SPI SPI_CLK_SEL (ssp5-clk-mux) is hogged and LED won't work.

Can we make the function selectable at run time other than not hogging SPI_CLK_SEL?

htot replied:
@andy-shev what to do here? Leave it for now?

htot replied:
If you decide we can leave it for now, I have tested leds.aml:
 
<details>
  <summary>Test results with/without spi enabled & with/without leds.aml loaded</summary>
**** This with ACPI_FEATURES_edison ?= "uart_2w i2c spi":

```
root@edison:~# blink-led 
Traceback (most recent call last):
  File "/usr/bin/blink-led", line 82, in <module>
    set_line("SPI_CLK_SEL", 0)
  File "/usr/bin/blink-led", line 66, in set_line
    line.request(consumer=line.owner().name(), type=gpiod.LINE_REQ_DIR_OUT)
OSError: [Errno 16] Device or resource busy
```

-> SPI_CLK_SEL is hogged by arduino.asl

With ACPI_FEATURES_edison ?= "uart_2w i2c"

```
root@edison:~# blink-led 
Traceback (most recent call last):
  File "/usr/bin/blink-led", line 86, in <module>
    led.request(consumer=led.owner().name(), type=gpiod.LINE_REQ_DIR_OUT)
OSError: [Errno 16] Device or resource busy
root@edison:~# modprobe -i ledtrig-heartbeat
```

-> GPIO40 hogged by leds.asl

```
modprobe -i ledtrig-heartbeat
echo 1 > /sys/class/leds/heartbeat/invert
```
-> LED blinks with heartbeat
```
rmdir /sys/kernel/config/acpi/table/leds 
```

-> does not release GPIO40

**** Booting without loading leds.aml

`blinkled 
`
blinks led by toggling GPIO40 from user space

</details>

andy-shev replied:
IIRC *leds.aml* is compatible with `gpio-leds` driver which serves platform device. To unload it properly we need my hack-patch. It seems it's not being shared.
Regarding to the run-time changing function — it's not possible for hogged pins.
I just realized that we might solve TRI_STATE_ALL issue whenever we will have support for ACPI r6.1 in pin control library. In that case we might write an ASL which contains pin function and configuration information. But it's not a nearest future.

htot replied:
So, for now I just need to document not to use the led in combination with spi - or reboot when switching function. Noted.

htot committed:
arduino.asl: create arduino-all.asl

Copy arduino.asl to arduino-all.asl which is a solution asl with
the following enabled:
MUX_I2C
MUX_SPI
MUX_UART_4WIRE

Remove defines from arduino.asl. This table requires to put the
defines on the iasl command line to make them under control of
the recipe or Makefile.

Signed-off-by: Ferry Toth <ftoth@exalondelft.nl>

htot commented:
I force pushed v3. I think all issues are resolved except toggle TRI_STATE_ALL. And one I opened on SPI_CLK_SEL. Not sure what to do about these.

htot commented:
I force pushed v4. There was an issue in acpi-tables-load where the switch to 'find ...' caused $TABLE to have absolute paths. This is fixed and has now been tested to work on edison.

andy-shev commented on recipes-bsp/acpica/acpica_20190215.bb:
```diff
@@ -17,9 +17,9 @@ COMPATIBLE_HOST = "(i.86|x86_64|arm|aarch64).*-linux"
-    "
-SRC_URI[md5sum] = "5aa086f71f4b5273c0932a1e04419a37"
-SRC_URI[sha256sum] = "d1e26cdde58938653a277a5ca59ce1df045508d345fee13fed2eaacc6ef51851"
+           "
```
Unrelated change.

htot replied:
The change in indent is accidental. The change in checksum due to upgrade to newer acpica.

andy-shev commented on recipes-bsp/acpi-tables/samples/edison/arduino.asl:
```diff
@@ -0,0 +1,27 @@
+/*
```
Please, fold this into previous patch.

htot replied:
Done

andy-shev commented on recipes-bsp/acpi-tables/acpi-tables.bb:
```diff
@@ -14,6 +14,14 @@ inherit deploy
 ACPI_TABLES[doc] = "List of ACPI tables to include with the initrd"
+ACPI_FEATURES_edison ?= "uart_2w spi i2c"
+IASLFLAGS = " \
+    ${@bb.utils.contains('ACPI_FEATURES', 'uart_2w', '-D MUX_UART_2WIRE', '', d)} \
```
If it works in a form `-DMUX_UART_2WIRE`  (no space) and so on for the rest, I would prefer it over spaceful -version.

htot replied:
Tested and done.

andy-shev commented on recipes-bsp/acpi-tables/samples/edison/leds.asli:
```diff
@@ -0,0 +1,85 @@
+            },
+            ToUUID("dbb8e3e6-5886-4ba6-8795-1319f52a966b"),
+            Package () {
+                Package () {"led-0", "\\_SB.LEDS.LED0"},
```
Simple `"LED0"` should be enough.

htot replied:
Indeed. Done

andy-shev commented on recipes-bsp/acpi-tables/acpi-tables.bb:
```diff
@@ -45,6 +52,24 @@ do_compile() {
+
+	install -d ${D}/kernel/firmware/acpi
+	for table in ${ACPI_TABLES}; do
+		dest_table=$(basename $table asl)
```
`asl` -> `.asl` and so for the following lines.

htot replied:
Done

andy-shev commented on recipes-bsp/acpi-tables/acpi-tables.bb:
```diff
@@ -10,6 +10,8 @@ LIC_FILES_CHKSUM = "file://${COREBASE}/meta/files/common-licenses/BSD;md5=377548
 
 DEPENDS = "acpica-native"
 
+FILESEXTRAPATHS_prepend := "${THISDIR}/files/${BPN}:"
```
Let's for now exclude this from the PR. I mean entire last patch.

htot replied:
Dropped

htot commented:
All comments should now be addressed.

htot commented:
Thanks!
